### PR TITLE
feat!: implement peer exchange

### PR DESCRIPTION
**Pull request by danisharora099**
## Problem

Currently, only _static_ peer discovery is available in js-waku, which limits the number of peers js-waku can discover.

To remediate to this, [34/WAKU2-PEER-EXCHANGE](https://rfc.vac.dev/spec/34/) was designed.

## Solution

Implement [Peer Exchange](https://rfc.vac.dev/spec/34/) as a protocol to discover peers for js-waku. 

## Notes

- ~~Currently blocked by: https://github.com/status-im/nwaku/issues/1408~~
- Resolves https://github.com/waku-org/js-waku/issues/910

### CODE REVIEW

**Comment by danisharora099** [comment]
Currently running into an error while decoding a message back from the node:
```
  error: Error: Failed to decode response
      at WakuPeerExchange.query (file:///Users/danisharora099/Developer/status/waku/packages/core/src/lib/waku_peer_exchange/index.ts:94:15)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async Context.<anonymous> (file:///Users/danisharora099/Developer/status/waku/packages/tests/tests/peer_exchange.node.spec.ts:37:29)
}```

File: packages/interfaces/src/index.ts
Comment (fryorcraken) [inline packages/interfaces/src/index.ts:59]: a `number` would be preferable

File: packages/interfaces/src/index.ts
Comment (fryorcraken) [inline packages/interfaces/src/index.ts:67]: This is not really useful per se. We do have a `ENR` class that could/should be used

Comment (fryorcraken) [reply packages/interfaces/src/index.ts:67]: Fair to merge with byte array and do follow up PR to provide `ENR` interface.

**Comment by danisharora099** [comment]
Blocked by: https://github.com/status-im/nwaku/issues/1408

File: packages/interfaces/src/index.ts
Comment (danisharora099) [reply packages/interfaces/src/index.ts:59]: Definitely. Will address this once the follow-up PRs are merged sequentially to avoid any divergent branches.

File: packages/core/src/proto/peer_exchange.proto
Comment (fryorcraken) [inline packages/core/src/proto/peer_exchange.proto:8]: The spec says `int`. I am guessing the `uint64` is coming from the nwaku implementation.

Would it make sense to use `uint32` instead @kaiserd? In JavaScript `uint64` means we use `Bigint` which should not really be necessary considering the functionality of this field.

File: packages/core/src/lib/waku.ts
Comment (fryorcraken) [inline packages/core/src/lib/waku.ts:190]:
```suggestion
    return  this.libp2p.dialProtocol(peer, codecs);
```

File: packages/core/src/lib/waku_peer_exchange/index.ts
Comment (fryorcraken) [inline packages/core/src/lib/waku_peer_exchange/index.ts:12]:
```suggestion
```

File: packages/interfaces/src/index.ts
Comment (fryorcraken) [inline packages/interfaces/src/index.ts:14]:
```suggestion
  PeerExchange = "peerexchange",
```

Let's keep same casing convention than light push (or you can change light push instead but be sure to update the changelog to note the breaking change).

File: packages/tests/src/nwaku.ts
Comment (fryorcraken) [inline packages/tests/src/nwaku.ts:59]: why this? Just add `discv5BootstrapNode` to the type.

File: packages/tests/src/nwaku.ts
Comment (fryorcraken) [inline packages/tests/src/nwaku.ts:454]: Yes I agree, just remove this default (you may have to change some tests).

```suggestion
```

File: packages/tests/package.json
Comment (fryorcraken) [inline packages/tests/package.json:51]:
```suggestion
```
I would suggest to configure your IDE instead.

File: packages/tests/tests/peer_exchange.node.spec.ts
Comment (fryorcraken) [inline packages/tests/tests/peer_exchange.node.spec.ts:40]: This is not needed. nwaku node automatically uses a free port and the harness automatically find it for you.
```suggestion
```

File: packages/tests/tests/peer_exchange.node.spec.ts
Comment (fryorcraken) [inline packages/tests/tests/peer_exchange.node.spec.ts:49]:
```suggestion
```

File: packages/tests/tests/peer_exchange.node.spec.ts
Comment (fryorcraken) [inline packages/tests/tests/peer_exchange.node.spec.ts:60]:
```suggestion
    const multiaddr = await nwaku2.getMultiaddrWithId();
```

File: packages/tests/tests/peer_exchange.node.spec.ts
Comment (fryorcraken) [inline packages/tests/tests/peer_exchange.node.spec.ts:54]: If these logs have any value when debugging issues with the CI (I don't think they do) then best to use the `debug` library.

```suggestion
   const log = debug("waku:test:peer-exchange"
   // ...
    log("created light node");
```

**Review by fryorcraken — APPROVED** [review]
Approved so you can move forward.

Only blocking comment in the additional `manualArgs`. Please just add any nwaku argument we need to use to the interface. Fine to fix in follow-up PR.

File: packages/core/src/proto/peer_exchange.proto
Comment (fryorcraken) [inline packages/core/src/proto/peer_exchange.proto:4]:
```suggestion
  optional bytes enr = 1;
```

camel case

File: packages/interfaces/src/index.ts
Comment (danisharora099) [reply packages/interfaces/src/index.ts:14]: The difference is Peer Exchange is defined as [`peer-exchange` in the protocol identifier in the specs],(https://rfc.vac.dev/spec/34/) and [Light Push as `lightpush`](https://rfc.vac.dev/spec/19/) -- so keeping it consistent with them?

File: packages/interfaces/src/index.ts
Comment (danisharora099) [reply packages/interfaces/src/index.ts]: addressed in the same branch

File: packages/tests/src/nwaku.ts
Comment (danisharora099) [reply packages/tests/src/nwaku.ts:448]: addressed this here: https://github.com/waku-org/js-waku/pull/1044

File: packages/peer-exchange/src/waku_peer_exchange_discovery.ts
Comment (fryorcraken) [inline packages/peer-exchange/src/waku_peer_exchange_discovery.ts:51]:
```suggestion
    const { protocols, peerId } = event.detail;
    if (!protocols.includes(PeerExchangeCodec)) return;
```
 
nipick: Just deconstruct all data from event upfront.

File: packages/peer-exchange/src/waku_peer_exchange_discovery.ts
Comment (fryorcraken) [inline packages/peer-exchange/src/waku_peer_exchange_discovery.ts:66]: Are you tracking the interval to re-query the same peer in an issue?

File: packages/peer-exchange/tsconfig.json
Comment (fryorcraken) [inline packages/peer-exchange/tsconfig.json:1]: Import from root tsconfig file

File: packages/tests/tests/peer_exchange.node.spec.ts
Comment (fryorcraken) [inline packages/tests/tests/peer_exchange.node.spec.ts:30]: If the test fails because we are still pending some investigation then just delete it, you can have a different branch that just introduces the test so that you and @kaiserd can investigate but no need to push it to master.

**Review by fryorcraken — APPROVED** [review]
empty file `packages/peer-exchange/.eslintignore`

LGTM.

Let's merge today please because #1052 conflicts on it

File: packages/peer-exchange/src/waku_peer_exchange_discovery.ts
Comment (danisharora099) [reply packages/peer-exchange/src/waku_peer_exchange_discovery.ts:66]: I'm not 100% sure that I understand the issue correctly.

Since there is an event listener for `change:protocols`, any time a new peer is discovered, it will automatically be run thorough by the callback function. 
I don't understand the need of running an interval and on what

**Comment by danisharora099** [comment]
Ready to be merged, once this is merged: https://github.com/waku-org/js-waku/pull/1057

Currently blocked because of circular dependencies between `core` and `peer-exchange` as `peer-exchange` is dependant on its `proto` definitions from `core`

cc @fryorcraken

File: packages/tests/tests/peer_exchange.node.spec.ts
Comment (fryorcraken) [inline packages/tests/tests/peer_exchange.node.spec.ts:18]:
```suggestion
  it("Test Fleet: Queries successfully [Live Data]", async function () {
```

File: packages/tests/tests/peer_exchange.node.spec.ts
Comment (fryorcraken) [inline packages/tests/tests/peer_exchange.node.spec.ts:27]:
```suggestion
    const waku = await createLightNode({
```

**Review by fryorcraken — APPROVED** [review]

**Pull request merged**
